documents/model.go

```diff
@@ -754,3 +755,154 @@ func isNFTInRole(role *coredocumentpb.Role, registry common.Address, tokenID []b
 
 	return false
 }
+
+// assembleAccessToken assembles a Read Access Token from the payload received
+func assembleAccessToken(dm *CoreDocumentModel, id identity.IDConfig, payload documentpb.AccessTokenParams) (*coredocumentpb.AccessToken, error) {
+	tokenIdentifier := utils.RandomSlice(32)
+	granterID := id.ID[:]
+	roleID := byte(len(dm.Document.Roles))
+	grantee, err := hexutil.Decode(payload.GetGrantee())
+	if err != nil {
+		return nil, err
+	}
+	docID, err := hexutil.Decode(payload.GetDocumentIdentifier())
+	if err != nil {
+		return nil, err
+	}
+	// assemble access token message to be signed
+	tm := append(tokenIdentifier, granterID...)
+	tm = append(tm, grantee...)
+	tm = append(tm, roleID)
+	tm = append(tm, docID...)
+
+	// fetch key pair from identity
+	privateKey := id.Keys[identity.KeyPurposeEthMsgAuth].PrivateKey
+	pubKey := id.Keys[identity.KeyPurposeEthMsgAuth].PublicKey
+
+	// sign the assembled access token message, return signature
+	sig, err := crypto.SignMessage(privateKey, tm, "CurveSecp256K1", true)
+	if err != nil {
+		return nil, err
+	}
+	// assemble the access token, appending the signature and public keys
+	at := new(coredocumentpb.AccessToken)
+	at.Identifier = tokenIdentifier
+	at.Granter = granterID
+	at.Grantee = grantee
+	at.RoleIdentifier = []byte{roleID}
+	at.DocumentIdentifier = docID
+	at.Signature = sig
+	at.Key = pubKey
+
+	return at, nil
+}
+
+// AddAccessTokenToReadRules adds the AccessToken(s) to the read rules of the document
+func (m *CoreDocumentModel) AddAccessTokenToReadRules(id identity.IDConfig, payload documentpb.AccessTokenParams) error {
+	at, err := assembleAccessToken(m, id, payload)
+	if err != nil {
+		return errors.New("failed to construct AT: %v", err)
+	}
+
+	role := new(coredocumentpb.Role)
+	rk, err := utils.ConvertIntToByte32(len(m.Document.Roles))
+	if err != nil {
+		return err
+	}
+	role.RoleKey = rk[:]
+	role.AccessTokens = append(role.AccessTokens, at)
+	m.addNewRule(role, coredocumentpb.Action_ACTION_READ)
+	err = m.setCoreDocumentSalts()
+	if err != nil {
+		return errors.New("failed to generate CoreDocumentSalts")
+	}
+	return nil
+}
+
+// ATOwnerCanRead checks if the owner of the AT can read the document
+func (m *CoreDocumentModel) ATOwnerCanRead(tokenReq p2ppb.AccessTokenRequest , account identity.CentID) error {
+	// check if the access token is present in read rules
+	token, err := m.findAT(coredocumentpb.Action_ACTION_READ, tokenReq.AccessTokenId, func(role *coredocumentpb.Role, tokenID []byte) (*coredocumentpb.AccessToken, error){
+		return isATInRole(role, tokenReq.AccessTokenId)
+	})
+	if err != nil {
+		return err
+	}
+
+	granterID, err := identity.ToCentID(token.Granter)
+	if err != nil {
+		return err
+	}
+	// check that the token granter has the correct access type for the document
+	validGranter := m.AccountCanRead(granterID)
+	if !validGranter {
```

Comment:
Remove `validGranter := m.AccountCanRead(granterID)` and do
```suggestion
	if m.AccountCanRead(granterID) {
```

Reply:
why is this preferable than !validGranter?

Reply:
just idiomatic go way of not creating unnecessary variables.